File: src/go/GenerateMatrix.go
package main

import "fmt"

func generateMatrix(n int) [][]int {
	left, right, top, bottom := 0, n-1, 0, n-1
	matrix := make([][]int,n)
	for i := 0; i < n; i++ {
		matrix[i] = make([]int, n)
	}
	max := n * n
	count := 1

	for i := count; i <= max; i++ {
		for i := left; i <= right; i++ {
			matrix[top][i] = count
			count++
		}
		top++

		for i := top; i <= bottom; i++ {
			matrix[i][right] = count
			count++
		}
		right--

		for i := right; i >= left; i-- {
			matrix[bottom][i] = count
			count++
		}
		bottom--

		for i := bottom; i >= top; i-- {
			matrix[i][left] = count
			count++
		}
		left++

	}
	return matrix
}

func main() {
	matrix := generateMatrix(5)
	for index, value := range matrix {
		fmt.Println(index,value)
	}
}

File: src/java/GenerateMatrix.java
package org.jy;

public class GenerateMatrix {

    public int[][] generateMatrix(int n) {
        int left = 0, right = n - 1, top = 0, bottom = n - 1;
        int[][] matrix = new int[n][n];
        int num = 1, tar = n * n;
        while (num <= tar) {
            for (int i = left; i <= right; i++) {
                matrix[top][i] = num++;
            } // left to right.
            top++;
            for (int i = top; i <= bottom; i++) {
                matrix[i][right] = num++;
            } // top to bottom.
            right--;
            for (int i = right; i >= left; i--) {
                matrix[bottom][i] = num++;
            } // right to left.
            bottom--;
            for (int i = bottom; i >= top; i--) {
                matrix[i][left] = num++;
            } // bottom to top.
            left++;
        }
        return matrix;
    }

    public static void main(String[] args) {
        int[][] ints = new GenerateMatrix().generateMatrix(5);
        for (int i = 0; i < ints.length; i++) {
            for (int y = 0; y < ints.length; y++) {
                System.out.print(ints[i][y] + " ");
            }
            System.out.println();
        }
    }

}
